### ISSUE TITLE
DP+TP prefill: last-token selection corrupts logits for batch rows past num_cores*TILE_HEIGHT

### ISSUE THREAD

**Issue by ssaliceTT** [tenstorrent/tt-xla#5929]
Under batched DP+TP chunked prefill, identical prompts placed in different batch slots produce different greedy output. The prefill hidden states are correct for every row, but the final logits are wrong for batch rows past a fixed boundary.

Root cause: select_hidden_states picks each request's last-token hidden with advanced indexing hidden_states[batch_indices, indices, :]. This lowers to a flat gather (flat = row*seq + pos) that returns the wrong row once the flat index reaches num_cores*TILE_HEIGHT (2048 on Wormhole, 64 cores times 32). Rows past 2048/seq_width read the wrong hidden state, so their logits and greedy tokens are wrong. It is independent of the transformer compute, the attention, and the KV cache, and reproduces in pure tensor parallel.

Repro: test_data_tensor_parallel_chunked_prefill_llmbox_large[Qwen/Qwen3-8B-64] (batch 64, seq width 128, first wrong row 16). The prefill hidden states match across duplicate prompts while the logits diverge from row 16.

Confirmed present on latest main.

Fix direction: select the last-token hidden with a one-hot matmul over the seq dim instead of the flat gather.

## IR-level detail

We dumped the IR for the original advanced-indexing version to confirm what actually runs.

What the frontend emits (StableHLO): a clean multi dimensional gather over the unflattened hidden states, not a flat gather.

    stablehlo.gather(%hidden, %idx)
      <collapsed_slice_dims=[0,1], start_index_map=[0,1], slice_sizes=[1,1,4096]>
      : (tensor<16x128x4096xbf16>, tensor<16x2xi32>) -> tensor<16x4096xbf16>

So the flattening is not a torch/xla behavior. It happens in the tt-mlir gather lowering. tt-metal has no general gather, so the gather is lowered to the embedding op (a row lookup into a table). The lowering reshapes the operand to a table of shape [batch*seq, hidden] and builds a single flat row index (flat = batch*seq_width + pos).

What actually runs (TTNN):

    ttnn.embedding(%idx, %table) : (tensor<16x1xui32>, tensor<2048x4096xbf16>) -> tensor<16x1x4096xbf16>

- table = tensor<2048x4096> = [batch*seq, hidden] = 16*128 x 4096, the flattened hidden states
- index = tensor<16x1>, one flat row id per request
- output = tensor<16x1x4096>, the 16 selected rows

For contrast, the input token embedding in the same graph is the same op with a real vocab table:

    ttnn.embedding(%tok, %weight) : (tensor<1024xui32>, tensor<151936x1024xbf16>) -> tensor<1024x1024xbf16>

## What is and is not the trigger

Table size is not the trigger. The vocab table has 151936 rows and its indices go up to 151935, and it works. So the kernel can address rows far past 2048.

Table layout is not the trigger. We checked both layouts. They are the same style:

    select table: (d0,d1)->(d0,d1), memref<2048x4096xbf16, dram>, interleaved
    vocab table:  (d0,d1)->(d0,d1), memref<151936x1024xbf16, dram>, interleaved

Both are plain row major, interleaved, identity mapped. The select table is not row sharded across a 64x32 grid. So the boundary is not coming from a special sharded layout.

The difference is the index provenance. The vocab index is the raw input token ids, genuine integers passed straight through. The select index is computed: flat = batch*seq_width + pos, built as an fp32 stride matmul and then cast to ui32. This computed index path is the suspect. See tt-mlir #9036 (force fp32 accumulation on index arithmetic matmuls feeding embedding).

Open question on the exact boundary. Pure index rounding does not cleanly predict 2048. bf16 loses integer exactness near 256 to 512, fp32 stays exact to about 16M, so neither rounds at 2048. The fact that the boundary is exactly num_cores*TILE_HEIGHT points at a reader or buffer limit inside the fused embedding kernel as well (the reader batches reads in tile height chunks). tt-metal #47506 is the same op and fused reader returning wrong rows on a large table, closed by adding synchronization barriers, at a different threshold. We could not localize our 2048 break to a single instruction, so this is index precision and/or a kernel reader limit, not a confirmed integer overflow.

Padding matters only indirectly. The flat index uses the padded chunk width as the stride, so more padding means a larger multiplier on the batch row, and the flat index crosses 2048 at a lower request number.

## Why the one-hot fix avoids all of this

bmm(onehot, hidden) lowers to a dot_general (matmul), not a gather, so the compiler never builds the flattened [batch*seq, hidden] table and never calls the embedding row lookup. It is correct for any batch, width, and position, with negligible cost (one small matmul per prefill).
